quickwit/quickwit-control-plane/src/ingest/ingest_controller.rs

```diff
@@ -0,0 +1,1348 @@
+
+    /// Pings an ingester to determine whether it is available for hosting a shard. If a follower ID
+    /// is provided, the leader candidate is in charge of pinging the follower candidate as
+    /// well.
+    async fn ping_leader(
+        &mut self,
+        ctx: &ActorContext<Self>,
+        leader_id: &NodeId,
+        follower_id_opt: Option<&NodeId>,
+    ) -> Result<(), PingLeaderError> {
+        let mut leader_ingester = self
+            .ingester_pool
+            .get(leader_id)
+            .await
+            .ok_or(PingLeaderError::LeaderUnavailable)?;
+
+        let ping_request = PingRequest {
+            leader_id: leader_id.clone().into(),
+            follower_id: follower_id_opt
+                .cloned()
+                .map(|follower_id| follower_id.into()),
+        };
+        ctx.protect_future(timeout(
+            PING_LEADER_TIMEOUT,
+            leader_ingester.ping(ping_request),
+        ))
+        .await
+        .map_err(|_| PingLeaderError::LeaderUnavailable)? // The leader timed out.
+        .map_err(|error| {
+            if let Some(follower_id) = follower_id_opt {
+                if matches!(error, IngestV2Error::IngesterUnavailable { ingester_id } if ingester_id == *follower_id) {
+                    return PingLeaderError::FollowerUnavailable;
+                }
+            }
+            PingLeaderError::LeaderUnavailable
+        })?;
+        Ok(())
+    }
+
+    /// Finds an available leader-follower pair to host a shard.
+    async fn find_leader(
+        &mut self,
+        ctx: &ActorContext<Self>,
+        unavailable_ingesters: &mut HashSet<NodeId>,
+    ) -> Option<(NodeId, Option<NodeId>)> {
+        let mut candidates: Vec<NodeId> = self.ingester_pool.keys().await;
+        candidates.shuffle(&mut rand::thread_rng());
+
+        #[cfg(test)]
+        candidates.sort();
+
+        if self.replication_factor == 1 {
+            for leader_id in candidates {
+                if unavailable_ingesters.contains(&leader_id) {
+                    continue;
+                }
+                if self.ping_leader(ctx, &leader_id, None).await.is_ok() {
+                    return Some((leader_id, None));
+                }
+            }
+        } else {
+            for (leader_id, follower_id) in candidates.into_iter().tuple_combinations() {
+                if unavailable_ingesters.contains(&leader_id)
+                    || unavailable_ingesters.contains(&follower_id)
+                {
+                    continue;
+                }
+                match self.ping_leader(ctx, &leader_id, Some(&follower_id)).await {
+                    Ok(_) => return Some((leader_id, Some(follower_id))),
+                    Err(PingLeaderError::LeaderUnavailable) => {
+                        unavailable_ingesters.insert(leader_id);
+                    }
+                    Err(PingLeaderError::FollowerUnavailable) => {
+                        unavailable_ingesters.insert(follower_id);
```

Comment:
I suggest we do not mark the `follower` as unavailable here.

```suggestion
                                      // We do not mark the `ingester` as unavailable here.
                                      // The idea here is that the problem could be specific to the network between the leader and this follower.
                                      // We'd better stick to defining unavailability as being unavailable from the control plane.
```